### PR TITLE
feat(module factory): add before_resolve hook

### PR DESCRIPTION
**Pull request by suxin2017**
## Related issue (if exists)

https://github.com/web-infra-dev/rspack/issues/2788
### <samp>🤖 Generated by Copilot at 22b4de9</samp>

Added a new plugin hook for normal module resolution in `rspack_core`. This hook allows plugins to modify or skip the default module resolution logic in `NormalModuleFactory`. The hook is implemented by the `Plugin` trait and the `PluginDriver` struct, and uses the `NormalModuleBeforeResolveArgs` struct to pass the relevant information.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 22b4de9</samp>

*  Add `before_resolve` hook to `Plugin` trait and `PluginDriver` struct to allow plugins to intercept and modify module resolution ([link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R91-R98), [link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR284-R296))
*  Modify `NormalModuleFactory` to invoke `before_resolve` hook before resolving module dependency using `ResolveArgs` struct ([link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-4e77827441656f6c0f2713ed92f4f4a58029006543ba5e686c842d1b28de6c0eL15-R16), [link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-4e77827441656f6c0f2713ed92f4f4a58029006543ba5e686c842d1b28de6c0eR88-R100), [link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eL14-R24))
*  Define and export `NormalModuleBeforeResolveArgs` struct in `plugin::args` and `plugin::api` to hold information about module dependency and context for `before_resolve` hook ([link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695L11-R11), [link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-65ed133c45bc1a8d647a0c4b50808420d3a9b9853e87ee835ca90995cbf5c337R97-R102))
*  Define and export `PluginNormalModuleFactoryBeforeResolveOutput` type alias in `plugin::api` to represent possible outcomes of `before_resolve` hook ([link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R27), [link](https://github.com/web-infra-dev/rspack/pull/2940/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eL14-R24))

</details>

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
Please rebase the main branch.

**Comment by suxin2017** [comment]
> Please rebase the main branch.
Why did I run `pnpm run build:cli:debug` after concatenation removed a lot of type definitions
image

**Comment by IWANABETHATGUY** [comment]
> > Please rebase the main branch.
> 
> Why did I run `pnpm run build:cli:debug` after concatenation removed a lot of type definitions image
> 
> @IWANABETHATGUY

Sorry, I can't reproduce.

**Comment by IWANABETHATGUY** [comment]
maybe you could try `build:cli:release`

**Comment by suxin2017** [comment]
> 

I don't know why this only happened to me, I didn't reproduce it on gitpod.io.

**Comment by IWANABETHATGUY** [comment]
Maybe related to https://github.com/web-infra-dev/rspack/issues/2912, I will investigate this issue when I have time.

**Comment by IWANABETHATGUY** [comment]
Would you mind resolving conflicts? Overall, this pull request looks good to me.

**Comment by suxin2017** [comment]
> 

sure

**Comment by suxin2017** [comment]
> Would you mind resolving conflicts? Overall, this pull request looks good to me.

I'm done

File: packages/rspack/tests/IgnorePlugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+const { IgnorePlugin } = require("webpack");
+
+class Plugin implements RspackPluginInstance {
+	name = "a";
+
+	apply(compiler: Compiler) {
+		// Wait for configuration preset plugions to apply all configure webpack defaults
+		compiler.hooks.compilation.tap("a", com => {
+			com.normalModuleFactory.hooks.beforeResolve.tap("a", (...args) => {
+				console.log(...args);
```
Comment (IWANABETHATGUY) [inline packages/rspack/tests/IgnorePlugin.test.ts:13]: console log in test is meaningless, if you want to  check something please use jest.expect or something similar.

File: Cargo.toml [hunk 2]
```diff
@@ -41,7 +41,7 @@ members = ["crates/*"]
   sugar_path         = { version = "0.0.12" }
   testing_macros     = { version = "0.2.9" }
   tokio              = { version = "1.28.0" }
-  tracing            = { version = "0.1.38" }
+  tracing            = { version = "0.1.37" }
```
Comment (IWANABETHATGUY) [inline Cargo.toml:44]: Any reason we need to downgrade the tracing crate?

**Comment by IWANABETHATGUY** [comment]
it looks ready to merge if you could address the rest of the two issues.

**Comment by suxin2017** [comment]
> it looks ready to merge if you could address the rest of the two issues.

this.#updateDisableHooks() like this https://github.com/web-infra-dev/rspack/pull/2940/files#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR417

I don't quite understand the problem. Everything else has been solved

**Comment by IWANABETHATGUY** [comment]
> Any reason we need to downgrade the tracing crate?
What about this one?

**Comment by suxin2017** [comment]
> > Any reason we need to downgrade the tracing crate?
> 
> What about this one?
> 

https://crates.io/crates/tracing/versions
This version 0.1.38 was removed

**Comment by IWANABETHATGUY** [comment]
> 

nice point, LGTM

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**